### DISCUSSION TITLE
Is there a better way to manually setError on a query?

### DISCUSSION THREAD

**Discussion by bruno-vinicius** [TanStack/query#1295] [Q&A] [answered]
I was messing around the query object and found a `dispatch` on it, so i make a 
`query.dispatch({ type: 3, error: { foo: "bar" } });` (`3` in this case is the enum value of ActionType.Error) 
and it worked. But i don't know if this is "clean enough".

**Comment by TkDodo** [comment]
can you elaborate a bit what your use-case would be? Why do you manually need to set it to error?

**Comment by bruno-vinicius** [comment, marked as answer]
It's a bit odd i know, but is a kind-of-authorization flow with the first request is given an error and the next request (this one that "resolves" the authorization) return the result of the first. So my solution was something like "hold" error state of the first request, make the second, and resolve or reject in case of a new error, the first one. I make this with another library (digging the code and use undocumented features) so was think in addressing this first before "full switch" to react-query.

**Reply by TkDodo** [reply]
> something like "hold" error state of the first request, make the second, and resolve or reject in case of a new error, the first one. 

Right, I think, you could just use one queryFn that does two requests and give this to `useQuery`. It will just be transparent for your queryCache, and you'll just have one cache entry. Something like:

```
const queryFn = async () => {
    let error
    let data
    try {
        const response = await axios.get('firstRequest')
        data = response.data
    } catch(e) {
        error = e
    } finally {
        try {
            const response = await axios.get(`secondRequest/${data ?? ''}`)
            return response.data
        }
        catch (e) {
            throw error ?? e
        }
    }
}

const { data, error, status } = useQuery('key', queryFn)
```

This looks a bit ugly and can surely be improved, code wise 😅 .
But what I'm doing here is:
- fire the firstRequest and capture the result / the error
- fire the secondRequest (`data` can be undefined, not sure how to build the request then, i just did empty string here 🤷 )
- if the second request also fails, we throw the first error
- if the second request succeeds, we get a resolved promise
- not sure what should happen if the first fails but the second one succeeds anyways, can that happen?

Anyways, I'm sure you get what I'm trying todo. React Query only expects you to have one function that returns a Promise, fulfilled or rejected. You can make as many requests as you want in that function.

Alternatively, you can set up two dependent queries and combine the error states somehow in a custom hook, but if the two requests are logically coupled together and only have "one" result together, I think this is "more correct". Hope that helps :)

**Reply by bruno-vinicius** [reply]
I didn't make myself clear, in between this two request, the user is redirected to a screen with a button and that fire the second request. So i think i cant chain the requests like you did. I think the "screen in the middle" of the two requests is the main reason that i need to "hold" and "re-resolve" or "re-reject" the first request imperatively.  Despite that, thanks for your explanation.

**Reply by TkDodo** [reply]
Yes, you did not mention that 😅 . Maybe you can post a codesandbox example that we can tinker with?

I think maybe the second suggestion of having two `useQuery`s  would work, something like:
- have one query and let it reject / resolve
- redirect the user to the second page, no matter what
- on that page, have a disabled query
- when the button is clicked, enable the query and let it run
- pass the data and error results from the first query to your second query
- resolve the second query with the error from the first

I think using the `enabled` prop to disable the second query until the user has clicked the button is a lot more declarative, but I would need to see something I can "play" with :)

**Reply by bruno-vinicius** [reply]
I made a very simple (and ugly) [codesandbox](https://codesandbox.io/s/react-query-set-error-2ysqf?file=/src/index.js) to play with.

**Reply by TkDodo** [reply]
what about something like this?

https://codesandbox.io/s/react-query-set-error-forked-v0pz2

**Reply by bruno-vinicius** [reply]
Good implementation, i think this will do the trick. Thanks a lot! 😄

**Reply by TkDodo** [reply]
The answer to many/most questions is “custom hooks” 😅

**Reply by bruno-vinicius** [reply]
Strong agree 😄 !!!

**Reply by bruno-vinicius** [reply]
I'm marking this thread as the answer but it puts my name on it instead of yours. 😢

**Reply by TkDodo** [reply]
Yes because you started the thread 🤷‍♂️. I don’t care as long as people will find the right answer👍

**Reply by bruno-vinicius** [reply]
I think i found a well-suited solution, i can wrap the base fetch on another Promise and pass the `resolve` and `reject` to be manually fired by another (in this case top-leve) place of my code (using Context). So i can make another request and resolves or reject the first one with a new data or error. Thanks for the help and the inspiration.

![image](https://user-images.githubusercontent.com/4923516/99730011-88992480-2a9a-11eb-8c03-b21dc74348c8.png)

**Comment by bruno-vinicius** [comment]
`queryCache.setQueryData` solve the "resolve" part, but i was thinking if was useful to have an error one.
